### PR TITLE
Change ClasspathUtils to support 'zip'-files implementation. Used Reflections API.

### PR DESCRIPTION
**Pull request by bcubk**
Fixed ClassPathUtils to support 'zip' files. I used the method 'safeClassForName' to initialize the retrieved classes. This PR is a fix of #1312
Issue: #1235

### CODE REVIEW

**Comment by timowest** [comment]
@bcubk Thanks a lot for the PR! Could you move ClasspathUtils to querydsl-codegen and also scope the needed dependencies to there?

**Comment by bcubk** [comment]
Yes, of course!

2015-04-09 22:22 GMT+02:00 Timo Westkämper <notifications@github.com>:

**Comment by timowest** [comment]
@Shredder121 @johnktims Does this PR look good to merge?

**Comment by Shredder121** [comment]
I'll check.

File: querydsl-codegen/src/test/java/com/querydsl/SomeOtherClass2.java [hunk 1]
```diff
@@ -1,4 +1,4 @@
-package com.querydsl.core.util;
+package com.querydsl;
```
Comment (johnktims) [inline querydsl-codegen/src/test/java/com/querydsl/SomeOtherClass2.java]: Shouldn't this be `com.querydsl.codegen`?

File: querydsl-codegen/src/main/java/com/querydsl/codegen/ClassPathUtils.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    public static Set<Class<?>> scanPackage(ClassLoader classLoader, Package pkg) throws IOException {
+        return scanPackage(classLoader, pkg.getName());
+    }
+
+    public static Set<Class<?>> scanPackage(ClassLoader classLoader, String pkg) throws IOException {
+        Reflections reflections = new Reflections(new ConfigurationBuilder()
+                .addUrls(ClasspathHelper.forPackage(pkg, classLoader))
+                .addClassLoader(classLoader)
+                .setScanners(new SubTypesScanner(false))
+                .setMetadataAdapter(new JavaReflectionAdapter()));
```
Comment (Shredder121) [inline querydsl-codegen/src/main/java/com/querydsl/codegen/ClassPathUtils.java]: [Looking at the implementation of `getMetadataAdapter()`](http://grepcode.com/file/repo1.maven.org/maven2/org.reflections/reflections/0.9.9/org/reflections/util/ConfigurationBuilder.java#ConfigurationBuilder.getMetadataAdapter%28%29) it is safe (and more extensible) to leave the construction of the `MetadataAdapter` to reflections.

Comment (bcubk) [inline querydsl-codegen/src/main/java/com/querydsl/codegen/ClassPathUtils.java]: Okay, so should I delete ```.setMetadataAdapter(new JavaReflectionAdapter()));```?

File: querydsl-codegen/src/test/java/com/querydsl/SomeOtherClass2.java (hunk 1 above)
Comment (bcubk) [inline querydsl-codegen/src/test/java/com/querydsl/SomeOtherClass2.java]: Yes, I'll move this to codeine package.

File: querydsl-codegen/src/main/java/com/querydsl/codegen/ClassPathUtils.java (hunk 2 above)
Comment (Shredder121) [inline querydsl-codegen/src/main/java/com/querydsl/codegen/ClassPathUtils.java]: Yes, I tested it and it works well.
This to enable extra adapters to be added by reflections. (unlikely, but the possibility is there now)

File: querydsl-codegen/src/test/java/com/querydsl/SomeOtherClass2.java (hunk 1 above)
Comment (Shredder121) [inline querydsl-codegen/src/test/java/com/querydsl/SomeOtherClass2.java]: Keep in mind that [`SomeOtherClass` and `SomeOtherClass2` depend on each other](https://github.com/querydsl/querydsl/pull/1313/files#diff-9fe16063e7692477327a759719d8c3a6R6).
The [test loads the `com.querydsl.codegen` package](https://github.com/querydsl/querydsl/pull/1313/files#diff-b8fb56070c211066760fb4bbc68ce563R39) and [verifies if the static block of `SomeOtherClass` was run](https://github.com/querydsl/querydsl/pull/1313/files#diff-b8fb56070c211066760fb4bbc68ce563R41).

File: querydsl-codegen/src/main/java/com/querydsl/codegen/ClassPathUtils.java (hunk 2 above)
Comment (bcubk) [inline querydsl-codegen/src/main/java/com/querydsl/codegen/ClassPathUtils.java]: Thank you for your feedback :+1:

**Comment by timowest** [comment]
@bcubk Thanks again for the contribution!

**Pull request merged**